keras/preprocessing/sequence.py

```diff
@@ -216,3 +218,103 @@ def _remove_long_seq(maxlen, seq, label):
+        start_index, end_index: delimiters for sub-sets of `data` & `targets`
+            to be used for batch generation. Useful to reserve part of
+            data for validation.
+        shuffle: whether to shuffle training samples, or draw them in
+            chronological order.
+        reverse: if `true`, input points are provided in reverse order.
+        batch_size: number of timeseries samples in each batch
+            (except maybe last one)
+
+    # Returns
+        A `Sequence` instance.
+
+    # Examples
+
+    ```data = np.array([[i] for i in range(50)])
+       targets = np.array([i for i in range(50)])
+
+       data_gen = TimeseriesGenerator(data, targets,
+                                      length=10, sampling_rate=2,
+                                      batch_size=2)
+       assert len(data_gen) == 20
+       assert (np.array_equal(data_gen[0][0],
+                              np.array([[[0], [2], [4], [6], [8]],
+                                        [[1], [3], [5], [7], [9]]])))
+       assert (np.array_equal(data_gen[0][1],
+                              np.array([[10], [11]])))
+    ```
+    """
+
+    def __init__(self, data, targets,
+                 length,
+                 sampling_rate=1,
+                 stride=1,
+                 start_index=0,
+                 end_index=None,
+                 shuffle=False,
+                 reverse=False,
+                 batch_size=128):
+        self.data = data
+        self.targets = targets
+        self.length = length
+        self.sampling_rate = sampling_rate
+        self.stride = stride
+        self.start_index = start_index + length
+        if end_index is None:
+            end_index = len(data) - 1
+        self.end_index = end_index
+        self.shuffle = shuffle
+        self.reverse = reverse
+        self.batch_size = batch_size
+
+    def __len__(self):
+        return int(np.ceil(
+            (self.end_index - self.start_index) /
+            (self.batch_size * self.stride)))
+
+    def __getitem__(self, index):
+        if self.shuffle:
+            rows = np.random.randint(
+                self.start_index, self.end_index, size=self.batch_size)
+        else:
+            i = self.start_index + self.batch_size * self.stride * index
+            rows = np.arange(i, min(i + self.batch_size *
+                                    self.stride, self.end_index), self.stride)
+
+        samples = np.empty((len(rows),
+                            self.length // self.sampling_rate,
+                            *self.data.shape[1:]))
+        targets = np.empty((len(rows), *self.targets.shape[1:]))
+        for j, row in enumerate(rows):
+            indices = range(rows[j] - self.length, rows[j], self.sampling_rate)
+            samples[j] = self.data[indices]
+            targets[j] = self.targets[rows[j]]
+        if self.reverse:
+            return samples[:, ::-1, :], targets
```

Comment:
This assumes a ndim of 3. The last : should be ...

Reply:
Aah thanks for pointing that out! Fixed.